[NVFuser] assert that vectors are the same size in translateSingleWelford

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #77010

Before, sometimes out_root.size() < in_root.size(), which would result
in a segfault while accessing out_root[i]. If, instead, we just error
out here, an exception will be thrown and then we'll run the fallback
instead of completely erroring out.

davidberard98 committed:
[NVFuser] assert that vectors are the same size in translateSingleWelford

Before, sometimes out_root.size() < in_root.size(), which would result
in a segfault while accessing out_root[i]. If, instead, we just error
out here, an exception will be thrown and then we'll run the fallback
instead of completely erroring out.

[ghstack-poisoned]

davidberard98 commented:
cc @jjsjann123 - in case we think the fix to the segfault in https://github.com/pytorch/pytorch/issues/76883 will take some time

jjsjann123 commented:
Ha good catch~ I'll try this with my repro~

But thanks for the assert. Regarding #76883, I'm gonna simply disable squeeze/unsqueeze. Let me push a PR for that~

jjsjann123 commented on torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp:
```diff
@@ -1930,6 +1930,7 @@ void TranslateApplicableWelford::translateSingleWelford(WelfordOp* welford) {
   //  largely taken from batchnorm cpp benchmark
   auto& in_root = in_val->getRootDomain();
   auto& out_root = out_avg->getRootDomain();
+  TORCH_INTERNAL_ASSERT(in_root.size() == out_root.size());
```
Naive comment, looking at the logic below, the check here should be `in_root.size() <= out_root.size()`?
```
for (const auto i : c10::irange(in_root.size())
   ... out_root[i] ...
```

davidberard98 replied:
sure, good point - I'll change that.

@jjsjann123 is this PR still useful now that the actual fixes are landed?

jjsjann123 replied:
Yes, let's merge this one.

davidberard98 committed:
Update on "[NVFuser] assert that vectors are the same size in translateSingleWelford"


Before, sometimes out_root.size() < in_root.size(), which would result
in a segfault while accessing out_root[i]. If, instead, we just error
out here, an exception will be thrown and then we'll run the fallback
instead of completely erroring out.

[ghstack-poisoned]

jjsjann123 reviewed (approved):
LGTM

davidberard98 commented:
@pytorchbot merge this please
